### PR TITLE
Evaluate models

### PR DESCRIPTION
**Pull request by ben-heil**
This PR creates the infrastructure that can be used to compare the performance of different classifiers and outlier detection models in predicting whether gene expression comes from a healthy or unhealthy sample.

The primary changes are
1. The creation of `evaluate_models.py`, a script that trains the models and evaluates them on a randomly sampled validation set
2. Modifying `classifier.py` to save the best model during training, allowing the best model to be used in `evaluate_models.py`

### CODE REVIEW

File: brdnet/classifier.py [hunk 1]
```diff
@@ -108,13 +108,20 @@ def prepare_input_data(Z_df, healthy_df, disease_df, random_seed):
...
-def load_data(args):
+def load_data(Z_file_path, healthy_file_path, disease_file_path, seed):
     '''Load and process the training data
 
     Arguments
     ---------
-    args: namespace
-        The command line arguments passed in to the script
+    Z_file_path: str or Path object
+        The path to the file containing the dataframe for moving expression data
```
Comment (ajlee21) [inline brdnet/classifier.py]: This may come from my lack of PLIER knowledge, but this description was a bit confusing to me.  Sounds like this is an output filename

File: brdnet/classifier.py [hunk 2]
```diff
@@ -128,34 +135,50 @@ def load_data(args):
     val_Y: numpy.array
         The labels for val_X
```
Comment (ajlee21) [inline brdnet/classifier.py]: I had to scroll up to figure out what "val_X" was here.  This might be a stylistic preference, but I think that repeating what the variables mean is a good idea.

File: brdnet/classifier.py [hunk 3]
```diff
@@ -128,34 +135,50 @@ def load_data(args):
...
     '''
-    Z_df = pandas.read_csv(args.Z_file_path, sep='\t')
+    Z_df = pandas.read_csv(Z_file_path, sep='\t')
 
     # Ensure the gene symbols are in alphabetical order
     Z_df = Z_df.sort_index()
```
Comment (ajlee21) [inline brdnet/classifier.py]: Just curious, why are you sorting here?

File: brdnet/classifier.py [hunk 4]
```diff
@@ -128,34 +135,50 @@ def load_data(args):
...
-    return prepare_input_data(Z_df, healthy_df, disease_df, args.seed)
+    return prepare_input_data(Z_df, healthy_df, disease_df, seed)
 
 
-def train_model(train_X, train_Y, val_X, val_Y,
-                model_name=None, logdir=None, lr=None, epochs=None, batch_size=16):
-    # Create log directory
-    os.mkdir(logdir)
-    model_name = model_name
+def train_model(train_X, train_Y, val_X, val_Y, checkpoint_path,
```
Comment (ajlee21) [inline brdnet/classifier.py:157]: Just curious, why did you choose those 3 models to start with?

File: brdnet/evaluate_models.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import pyod
+import sklearn
+
+import classifier
+import models
+import utils
+
+
+def model_from_pyod(model_name):
+    try:
```
Comment (ajlee21) [inline brdnet/evaluate_models.py:36]: Very minor, but looks like you're missing the description for the functions in evaluate_models.py

**Review by ajlee21 — APPROVED** [review]
Very nice PR!  I'm really interested to learn more about your project.  Some minor comments to make these files a bit clearer and then some comments for my own curiosity.

I'm curious what dataset you're using - how are cases and controls defined in this case?  How well did these model perform.  Something we can discuss offline.

One more comment on a previous commit that I couldn't mark in line:  What is "LV space" in reduce_dimensionality function?

File: brdnet/classifier.py (hunk 4 above)
Comment (ben-heil) [reply brdnet/classifier.py:157]: I wanted to see how performance would change with progressively more complicated models, so I started with Logistic Regression, scaled up to a small multilayer perceptron, then tried a similar ten-layer network

File: brdnet/classifier.py (hunk 3 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Since the script [uses matrix multiplication to transform the training data from gene space to PLIER space](https://github.com/greenelab/brd-net/blob/4831abb34632fabec88a0cdd3d3b747429060763/brdnet/classifier.py#L32), I need to make sure that the genes in `Z_df` and in `train_X/train_Y/val_X/val_Y` are in the same order. 

I get why this is confusing though. I had this call to `sort_index` outside of the `reduce_dimensionality` function for efficiency's sake, but I'll add some code to call it in the function efficiently.

File: brdnet/evaluate_models.py (hunk 5 above)
Comment (ben-heil) [reply brdnet/evaluate_models.py:36]: Good catch!

File: brdnet/classifier.py (hunk 1 above)
Comment (ben-heil) [reply brdnet/classifier.py]: Does it sound better if I change it from 'moving expression data into the PLIER latent space' to 'transforming expression data into a low dimensional representation'?
